### PR TITLE
wicked: Fix duplicate mac-addr errors in wicked-testsuite

### PR DESCRIPTION
**Pull request by cfconrad**
For some tests, e.g. macvtap or bridge-with-dummy, we use predefined
fixed mac-addresses.
If the order of the interfaces is such, that the bridge gets the
mac-address of the dummy-interface, while we have the same job parallel
running on a other multi-machine setup, the backend ovs-switch has two
times the same mac-address. The result are unexpected network errors.

- Related ticket: 
  - https://gitlab.suse.de/wicked-maintainers/wicked/-/issues/111
  - https://gitlab.suse.de/wicked-maintainers/wicked/-/issues/176
## Verification run:
  - https://openqa.suse.de/tests/8425126

### In wicked-ci
```
[ 2703s] Target: SLE_12_SP5@x86_64
[ 2703s] OpenQA URL: http://openqa.wicked.suse.de/tests/overview?distri=sle&version=12-SP5&build=No-Branch
[ 2703s] Name                      State    URL
[ 2703s] -----------------------   ------   -----------------------------------
[ 2703s] Name                      State    URL
[ 2703s] ----                      -----    ---
[ 2703s] wicked_advanced_ref       passed   http://openqa.wicked.suse.de/t53761
[ 2703s] wicked_advanced_sut       passed   http://openqa.wicked.suse.de/t53762
[ 2703s] wicked_aggregate_ref      passed   http://openqa.wicked.suse.de/t53763
[ 2703s] wicked_aggregate_sut      passed   http://openqa.wicked.suse.de/t53764
[ 2703s] wicked_basic_ref          passed   http://openqa.wicked.suse.de/t53765
[ 2703s] wicked_basic_sut          passed   http://openqa.wicked.suse.de/t53766
[ 2703s] wicked_ipv6_ref           passed   http://openqa.wicked.suse.de/t53767
[ 2703s] wicked_ipv6_sut           passed   http://openqa.wicked.suse.de/t53768
[ 2703s] wicked_startandstop_ref   passed   http://openqa.wicked.suse.de/t53769
[ 2703s] wicked_startandstop_sut   passed   http://openqa.wicked.suse.de/t53770
[ 2703s] wicked_sysctl_ref         passed   http://openqa.wicked.suse.de/t53771
[ 2703s] wicked_sysctl_sut         failed   http://openqa.wicked.suse.de/t53772
[ 2703s] wicked_wlan_ref           passed   http://openqa.wicked.suse.de/t53773
[ 2703s] wicked_wlan_sut           passed   http://openqa.wicked.suse.de/t53774
[ 2703s]
[ 2703s] Target: SLE_15_SP3@x86_64
[ 2703s] OpenQA URL: http://openqa.wicked.suse.de/tests/overview?distri=sle&version=15-SP3&build=No-Branch
[ 2703s] Name                      State    URL
[ 2703s] -----------------------   ------   -----------------------------------
[ 2703s] Name                      State    URL
[ 2703s] ----                      -----    ---
[ 2703s] wicked_advanced_ref       passed   http://openqa.wicked.suse.de/t53775
[ 2703s] wicked_advanced_sut       passed   http://openqa.wicked.suse.de/t53776
[ 2703s] wicked_aggregate_ref      passed   http://openqa.wicked.suse.de/t53777
[ 2703s] wicked_aggregate_sut      passed   http://openqa.wicked.suse.de/t53778
[ 2703s] wicked_basic_ref          passed   http://openqa.wicked.suse.de/t53779
[ 2703s] wicked_basic_sut          passed   http://openqa.wicked.suse.de/t53780
[ 2703s] wicked_ipv6_ref           passed   http://openqa.wicked.suse.de/t53781
[ 2703s] wicked_ipv6_sut           passed   http://openqa.wicked.suse.de/t53782
[ 2703s] wicked_startandstop_ref   passed   http://openqa.wicked.suse.de/t53783
[ 2703s] wicked_startandstop_sut   passed   http://openqa.wicked.suse.de/t53784
[ 2703s] wicked_sysctl_ref         passed   http://openqa.wicked.suse.de/t53785
[ 2703s] wicked_sysctl_sut         passed   http://openqa.wicked.suse.de/t53786
[ 2703s] wicked_wlan_ref           passed   http://openqa.wicked.suse.de/t53787
[ 2703s] wicked_wlan_sut           passed   http://openqa.wicked.suse.de/t53788
[ 2703s]
[ 2703s] Target: SLE_15_SP4@x86_64
[ 2703s] OpenQA URL: http://openqa.wicked.suse.de/tests/overview?distri=sle&version=15-SP4&build=No-Branch
[ 2703s] Name                      State    URL
[ 2703s] -----------------------   ------   -----------------------------------
[ 2703s] Name                      State    URL
[ 2703s] ----                      -----    ---
[ 2703s] wicked_advanced_ref       passed   http://openqa.wicked.suse.de/t53789
[ 2703s] wicked_advanced_sut       passed   http://openqa.wicked.suse.de/t53790
[ 2703s] wicked_aggregate_ref      passed   http://openqa.wicked.suse.de/t53791
[ 2703s] wicked_aggregate_sut      passed   http://openqa.wicked.suse.de/t53792
[ 2703s] wicked_basic_ref          passed   http://openqa.wicked.suse.de/t53793
[ 2703s] wicked_basic_sut          passed   http://openqa.wicked.suse.de/t53794
[ 2703s] wicked_ipv6_ref           passed   http://openqa.wicked.suse.de/t53795
[ 2703s] wicked_ipv6_sut           passed   http://openqa.wicked.suse.de/t53796
[ 2703s] wicked_startandstop_ref   passed   http://openqa.wicked.suse.de/t53797
[ 2703s] wicked_startandstop_sut   passed   http://openqa.wicked.suse.de/t53798
[ 2703s] wicked_sysctl_ref         passed   http://openqa.wicked.suse.de/t53799
[ 2703s] wicked_sysctl_sut         passed   http://openqa.wicked.suse.de/t53800
[ 2703s] wicked_wlan_ref           passed   http://openqa.wicked.suse.de/t53801
[ 2703s] wicked_wlan_sut           passed   http://openqa.wicked.suse.de/t53802
[ 2703s]
[ 2703s] Target: openSUSE_Leap_15.3@x86_64
[ 2703s] OpenQA URL: http://openqa.wicked.suse.de/tests/overview?distri=opensuse&version=15.3&build=No-Branch
[ 2703s] Name                      State    URL
[ 2703s] -----------------------   ------   -----------------------------------
[ 2703s] Name                      State    URL
[ 2703s] ----                      -----    ---
[ 2703s] wicked_advanced_ref       passed   http://openqa.wicked.suse.de/t53803
[ 2703s] wicked_advanced_sut       passed   http://openqa.wicked.suse.de/t53804
[ 2703s] wicked_aggregate_ref      passed   http://openqa.wicked.suse.de/t53805
[ 2703s] wicked_aggregate_sut      passed   http://openqa.wicked.suse.de/t53806
[ 2703s] wicked_basic_ref          passed   http://openqa.wicked.suse.de/t53807
[ 2703s] wicked_basic_sut          passed   http://openqa.wicked.suse.de/t53808
[ 2703s] wicked_ipv6_ref           passed   http://openqa.wicked.suse.de/t53809
[ 2703s] wicked_ipv6_sut           passed   http://openqa.wicked.suse.de/t53810
[ 2703s] wicked_startandstop_ref   passed   http://openqa.wicked.suse.de/t53811
[ 2703s] wicked_startandstop_sut   passed   http://openqa.wicked.suse.de/t53812
[ 2703s] wicked_sysctl_ref         passed   http://openqa.wicked.suse.de/t53813
[ 2703s] wicked_sysctl_sut         passed   http://openqa.wicked.suse.de/t53814
[ 2703s] wicked_wlan_ref           passed   http://openqa.wicked.suse.de/t53815
[ 2703s] wicked_wlan_sut           passed   http://openqa.wicked.suse.de/t53816
[ 2703s]
[ 2703s] Target: openSUSE_Tumbleweed@i586
[ 2703s] OpenQA URL: http://openqa.wicked.suse.de/tests/overview?distri=opensuse&version=Tumbleweed&build=No-Branch
[ 2703s] Name                      State    URL
[ 2703s] -----------------------   ------   -----------------------------------
[ 2703s] Name                      State    URL
[ 2703s] ----                      -----    ---
[ 2703s] wicked_advanced_ref       passed   http://openqa.wicked.suse.de/t53817
[ 2703s] wicked_advanced_sut       passed   http://openqa.wicked.suse.de/t53818
[ 2703s] wicked_aggregate_ref      passed   http://openqa.wicked.suse.de/t53819
[ 2703s] wicked_aggregate_sut      passed   http://openqa.wicked.suse.de/t53820
[ 2703s] wicked_basic_ref          passed   http://openqa.wicked.suse.de/t53821
[ 2703s] wicked_basic_sut          passed   http://openqa.wicked.suse.de/t53822
[ 2703s] wicked_ipv6_ref           passed   http://openqa.wicked.suse.de/t53823
[ 2703s] wicked_ipv6_sut           passed   http://openqa.wicked.suse.de/t53824
[ 2703s] wicked_startandstop_ref   passed   http://openqa.wicked.suse.de/t53825
[ 2703s] wicked_startandstop_sut   passed   http://openqa.wicked.suse.de/t53826
[ 2703s] wicked_sysctl_ref         passed   http://openqa.wicked.suse.de/t53827
[ 2703s] wicked_sysctl_sut         passed   http://openqa.wicked.suse.de/t53828
[ 2703s] wicked_wlan_ref           passed   http://openqa.wicked.suse.de/t53829
[ 2703s] wicked_wlan_sut           passed   http://openqa.wicked.suse.de/t53830
[ 2703s]
[ 2703s] Target: openSUSE_Tumbleweed@x86_64
[ 2703s] OpenQA URL: http://openqa.wicked.suse.de/tests/overview?distri=opensuse&version=Tumbleweed&build=No-Branch
[ 2703s] Name                      State    URL
[ 2703s] -----------------------   ------   -----------------------------------
[ 2703s] Name                      State    URL
[ 2703s] ----                      -----    ---
[ 2703s] wicked_advanced_ref       passed   http://openqa.wicked.suse.de/t53831
[ 2703s] wicked_advanced_sut       passed   http://openqa.wicked.suse.de/t53832
[ 2703s] wicked_aggregate_ref      passed   http://openqa.wicked.suse.de/t53833
[ 2703s] wicked_aggregate_sut      passed   http://openqa.wicked.suse.de/t53834
[ 2703s] wicked_basic_ref          passed   http://openqa.wicked.suse.de/t53835
[ 2703s] wicked_basic_sut          passed   http://openqa.wicked.suse.de/t53836
[ 2703s] wicked_ipv6_ref           passed   http://openqa.wicked.suse.de/t53837
[ 2703s] wicked_ipv6_sut           passed   http://openqa.wicked.suse.de/t53838
[ 2703s] wicked_startandstop_ref   passed   http://openqa.wicked.suse.de/t53839
[ 2703s] wicked_startandstop_sut   passed   http://openqa.wicked.suse.de/t53840
[ 2703s] wicked_sysctl_ref         passed   http://openqa.wicked.suse.de/t53841
[ 2703s] wicked_sysctl_sut         passed   http://openqa.wicked.suse.de/t53842
[ 2703s] wicked_wlan_ref           passed   http://openqa.wicked.suse.de/t53843
[ 2703s] wicked_wlan_sut           passed   http://openqa.wicked.suse.de/t53844
```

The failing http://openqa.wicked.suse.de/t53772 from SLE-12-SP5 looks like something different. I run a rerun on master and the error was still there http://openqa.wicked.suse.de/tests/53850 .

### CODE REVIEW

File: data/wicked/check_macvtap.c [hunk 1]
```diff
@@ -45,6 +46,24 @@ char fn[FILENAME_LENGTH];
 // Read buffer
 unsigned char buffer[PACKET_SIZE];
 
+void read_macaddr() {
+#define SYSFS_MAC "/sys/class/net/macvtap1/address"
+  unsigned char *mac = (unsigned char*) &arp_packet[DEST_MAC1_POS_IN_ARRAY];
+
+  FILE *fp = fopen(SYSFS_MAC, "r");
+  if (fp == NULL){
+    perror("Cannot open " SYSFS_MAC);
```
Comment (asmorodskyi) [inline data/wicked/check_macvtap.c:55]: mmm , I was not writing C code for ages but I assume that is not how string concatenation works in C ...

File: data/wicked/check_macvtap.c [hunk 2]
```diff
@@ -45,6 +46,24 @@ char fn[FILENAME_LENGTH];
 // Read buffer
 unsigned char buffer[PACKET_SIZE];
 
+void read_macaddr() {
```
Comment (asmorodskyi) [inline data/wicked/check_macvtap.c:49]: isn't there something in stdlib knowing how to read MAC ? or in kernel syscalls ? looks like we inventing the wheel here ...

File: lib/wickedbase.pm [hunk 3]
```diff
@@ -363,6 +363,26 @@ sub create_tunnel_with_commands {
...
+
+    my $prefix = $args{prefix} // 'BA:00';
+    $prefix =~ s/:/_/;
+    $prefix = hex($prefix);
+
+    $self->{unique_macaddr_cnt} = $self->{unique_macaddr_cnt} ? 0 : $self->{unique_macaddr_cnt} + 1;
+    $prefix += $self->{unique_macaddr_cnt};
+
+    my $w_id = get_required_var('WORKER_ID');
+    die("WORKER_ID to big!") if ($w_id > 0xffffffff);
```
Comment (asmorodskyi) [inline lib/wickedbase.pm:377]:
```suggestion
    die("WORKER_ID too big!") if ($w_id > 0xffffffff);
```

File: lib/wickedbase.pm [hunk 4]
```diff
@@ -363,6 +363,26 @@ sub create_tunnel_with_commands {
     assert_script_run("ip addr");
 }
 
+sub unique_macaddr {
```
Comment (asmorodskyi) [inline lib/wickedbase.pm:366]: there are several issues with this function : 
1. I am not sure that such fully random approach will not bring corner cases due to IPv6 specification . AFAIK IPv6 have some IP ranges  which have different purpose and will behave differently. So after all test which using this generator may produce different results based on output of this function. So we probably need not fully blind random here but something more intelligent taking IPv6 specification into account 
2. I assume there should be some perl/ C libraries which doing that smart 
3. artificial limitation which not allow running test on workers with ID greater than X is *NO GO* . Keep in mind that wicked tests are MM and MM tests due to implementation of tun/tap logic getting workers with ID + 64 . But I assume this one will be auto-resolved if we will process first two

File: data/wicked/check_macvtap.c (hunk 2 above)
Comment (cfconrad) [reply data/wicked/check_macvtap.c:49]: I guess there might be something in netlink, but I expect it is much more difficult then asking sysfs.

File: data/wicked/check_macvtap.c (hunk 1 above)
Comment (cfconrad) [reply data/wicked/check_macvtap.c:55]: It is, if you have two strings, you can just concatinate them. I think it is made by the preprocessor, so it is done during compile time, *not* runtime.

Comment (cfconrad) [reply data/wicked/check_macvtap.c:55]: see here: https://onlinegdb.com/NLe9ls3cI
